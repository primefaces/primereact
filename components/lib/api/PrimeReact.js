import { FilterMatchMode } from './FilterMatchMode';

export default class PrimeReact {
    static ripple = false;

    static inputStyle = 'outlined';

    static locale = 'en';

    static appendTo = null;

    static cssTransition = true;

    static autoZIndex = true;

    static nonce = null;

    static nullSortOrder = 1;

    static zIndex = {
        modal: 1100,
        overlay: 1000,
        menu: 1000,
        tooltip: 1100,
        toast: 1200
    };

    static filterMatchModeOptions = {
        text: [FilterMatchMode.STARTS_WITH, FilterMatchMode.CONTAINS, FilterMatchMode.NOT_CONTAINS, FilterMatchMode.ENDS_WITH, FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS],
        numeric: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO],
        date: [FilterMatchMode.DATE_IS, FilterMatchMode.DATE_IS_NOT, FilterMatchMode.DATE_BEFORE, FilterMatchMode.DATE_AFTER]
    };
}
