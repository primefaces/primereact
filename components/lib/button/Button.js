import * as React from 'react';
import { Ripple } from '../ripple/Ripple';
import { Tooltip } from '../tooltip/Tooltip';
import { classNames, IconUtils, ObjectUtils } from '../utils/Utils';
import { ButtonBase } from './ButtonBase';

export const Button = React.memo(
    React.forwardRef((inProps, ref) => {
        const props = ButtonBase.getProps(inProps);

        const elementRef = React.useRef(ref);

        React.useEffect(() => {
            ObjectUtils.combinedRefs(elementRef, ref);
        }, [elementRef, ref]);

        if (props.visible === false) {
            return null;
        }

        const createIcon = () => {
            const icon = props.loading ? props.loadingIcon : props.icon;
            const className = classNames('p-button-icon p-c', {
                'p-button-loading-icon': props.loading,
                [`p-button-icon-${props.iconPos}`]: props.label
            });

            return IconUtils.getJSXIcon(icon, { className }, { props });
        };

        const createLabel = () => {
            if (props.label) {
                return <span className="p-button-label p-c">{props.label}</span>;
            }

            return !props.children && !props.label && <span className="p-button-label p-c" dangerouslySetInnerHTML={{ __html: '&nbsp;' }}></span>;
        };

        const createBadge = () => {
            if (props.badge) {
                const badgeClassName = classNames('p-badge', props.badgeClassName);

                return <span className={badgeClassName}>{props.badge}</span>;
            }

            return null;
        };

        const disabled = props.disabled || props.loading;
        const showTooltip = !disabled || (props.tooltipOptions && props.tooltipOptions.showOnDisabled);
        const hasTooltip = ObjectUtils.isNotEmpty(props.tooltip) && showTooltip;
        const otherProps = ButtonBase.getOtherProps(props);
        const className = classNames('p-button p-component', props.className, {
            'p-button-icon-only': (props.icon || (props.loading && props.loadingIcon)) && !props.label && !props.children,
            'p-button-vertical': (props.iconPos === 'top' || props.iconPos === 'bottom') && props.label,
            'p-disabled': disabled,
            'p-button-loading': props.loading,
            'p-button-loading-label-only': props.loading && !props.icon && props.label,
            [`p-button-loading-${props.iconPos}`]: props.loading && props.loadingIcon && props.label
        });

        const icon = createIcon();
        const label = createLabel();
        const badge = createBadge();
        const defaultAriaLabel = props.label ? props.label + (props.badge ? ' ' + props.badge : '') : props['aria-label'];

        return (
            <>
                <button ref={elementRef} aria-label={defaultAriaLabel} {...otherProps} className={className} disabled={disabled}>
                    {icon}
                    {label}
                    {props.children}
                    {badge}
                    <Ripple />
                </button>
                {hasTooltip && <Tooltip target={elementRef} content={props.tooltip} {...props.tooltipOptions} />}
            </>
        );
    })
);

Button.displayName = 'Button';
