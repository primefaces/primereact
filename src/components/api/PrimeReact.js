export default class PrimeReact {

    static ripple = false;

    static locale = 'en';

    static zIndex = {
        modal: 1100,
        overlay: 1000,
        menu: 1000,
        tooltip: 1100
    }
}
