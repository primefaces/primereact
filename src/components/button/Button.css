/* Button */
.p-button {
    display: inline-block;
    position: relative;
    padding: 0;
    text-decoration: none !important;
    cursor: pointer;
    text-align: center;
    zoom: 1;
    overflow: visible; /* the overflow property removes extra width in IE */
    margin-right: 0.25em;
    user-select: none;
    -moz-user-select: none;
    -webkit-user-select: none;
}

/*button text element */
.p-button .p-button-text {
    display: block;
    line-height: normal;
}

.p-button-text-only .p-button-text {
    padding: .25em 1em;
}

.p-button-icon-only .p-button-text,
.p-button-text-empty .p-button-text {
    padding: .25em;
    text-indent: -9999999px;
}

.p-button-text-icon-left .p-button-text {
    padding: .25em 1em .25em 2.1em;
}

.p-button-text-icon-right .p-button-text {
    padding: .25em 2.1em .25em 1em;
}

/*button icon element(s) */
.p-button-icon-only .p-button-icon-left,
.p-button-text-icon-left .p-button-icon-left,
.p-button-text-icon-right .p-button-icon-right {
    position: absolute;
    top: 50%;
    margin-top: -.5em;
    height: 1em;
}

.p-button-icon-only .p-button-icon-left {
    top: 50%;
    left: 50%;
    margin-top: -.5em;
    margin-left: -.5em;
    width: 1em;
    height: 1em;
}

.p-button-icon-left {
    left: .5em;
}

.p-button-icon-right {
    right: .5em;
}

/*button sets*/
.p-buttonset .p-button {
    margin-left: 0;
    margin-right: 0;
}

/* workarounds */
button.p-button::-moz-focus-inner {
    border: 0; padding: 0; /* reset extra padding in Firefox */
}

/** Fluid **/
.p-fluid .p-button {
    width: 100%;
}

.p-fluid .p-button-text-icon-left .p-button-text,
.p-fluid .p-button-text-icon-right .p-button-text {
    padding-left: 1em;
    padding-right: 1em;
}

/** ButtonSet **/
.p-fluid .p-buttonset {
    width: 100%;
}

.p-fluid .p-buttonset.p-buttonset-1 .p-button {width: 100%;}
.p-fluid .p-buttonset.p-buttonset-2 .p-button {width: 50%;}
.p-fluid .p-buttonset.p-buttonset-3 .p-button {width: 33.3%;}
.p-fluid .p-buttonset.p-buttonset-4 .p-button {width: 25%;}
.p-fluid .p-buttonset.p-buttonset-5 .p-button {width: 20%;}
.p-fluid .p-buttonset.p-buttonset-6 .p-button {width: 16.6%;}

@media (max-width: 640px) {
    .p-fluid .p-buttonset.p-buttonset-1 .p-button,
    .p-fluid .p-buttonset.p-buttonset-2 .p-button,
    .p-fluid .p-buttonset.p-buttonset-3 .p-button,
    .p-fluid .p-buttonset.p-buttonset-4 .p-button,
    .p-fluid .p-buttonset.p-buttonset-5 .p-button,
    .p-fluid .p-buttonset.p-buttonset-6 .p-button {
        width: 100%;
    }
}

/** Outlined **/
.p-button.p-button-outlined {
    box-sizing: border-box;
    -moz-box-sizing: border-box;
    -webkit-box-sizing: border-box;
    background-color: transparent !important;
}
