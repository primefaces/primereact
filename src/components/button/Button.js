import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils/ClassNames';
import { tip } from '../tooltip/Tooltip';
import ObjectUtils from '../utils/ObjectUtils';
import { Ripple } from '../ripple/Ripple';

export class ButtonComponent extends Component {

    static defaultProps = {
        label: null,
        icon: null,
        iconPos: 'left',
        badge: null,
        badgeClassName: null,
        tooltip: null,
        tooltipOptions: null,
        forwardRef: null,
        disabled: false,
        loading: false,
        loadingOptions: null
    }

    static propTypes = {
        label: PropTypes.string,
        icon: PropTypes.any,
        iconPos: PropTypes.string,
        badge: PropTypes.string,
        badgeClassName: PropTypes.string,
        tooltip: PropTypes.string,
        tooltipOptions: PropTypes.object,
        forwardRef: PropTypes.any,
        disabled: PropTypes.bool,
        loading: PropTypes.bool,
        loadingOptions: PropTypes.any
    };

    getElementRef(el) {
        this.element = el;

        if (this.props.forwardRef) {
            if (ObjectUtils.isFunction(this.props.forwardRef)) {
                return this.props.forwardRef(el);
            }
            else {
                return this.props.forwardRef;
            }
        }

        return this.element;
    }

    componentDidMount() {
        if (this.props.tooltip) {
            this.renderTooltip();
        }
    }

    componentDidUpdate(prevProps) {
        if (prevProps.tooltip !== this.props.tooltip || prevProps.tooltipOptions !== this.props.tooltipOptions) {
            if (this.tooltip)
                this.tooltip.update({ content: this.props.tooltip, ...(this.props.tooltipOptions || {}) });
            else
                this.renderTooltip();
        }
    }

    componentWillUnmount() {
        if (this.tooltip) {
            this.tooltip.destroy();
            this.tooltip = null;
        }
    }

    getOptions(loadingOptions) {
        let disabled = this.isDisabled(loadingOptions);
        let vertical = this.isVertical(loadingOptions);
        let iconOnly = this.isIconOnly(loadingOptions);

        return { disabled, vertical, iconOnly };
    }

    getLoadingOptions() {
        if (this.props.loading) {
            let defaultOptions = { icon: 'pi pi-spin pi-spinner', position: 'left', disabled: true };
            let options = { ...defaultOptions, ...(this.props.loadingOptions || {}) };

            return options;
        }

        return null;
    }

    isDisabled(loadingOptions) {
        return this.props.disabled || (loadingOptions && loadingOptions.disabled);
    }

    isIconOnly(loadingOptions) {
        return !this.props.label && (this.props.icon || (loadingOptions && loadingOptions.icon))
    }

    isVertical(loadingOptions) {
        return this.props.label && (
            (this.props.iconPos === 'top' || this.props.iconPos === 'bottom') ||
            (loadingOptions && (loadingOptions.position === 'top' || loadingOptions.position === 'bottom'))
        );
    }

    renderTooltip() {
        this.tooltip = tip({
            target: this.element,
            content: this.props.tooltip,
            options: this.props.tooltipOptions
        });
    }

    renderIcon(icon, position) {
        let content = null;

        if (icon) {
            let iconType = typeof icon;
            let className = classNames('p-c', {
                [`${icon}`]: iconType === 'string',
                'p-button-icon-left': position === 'left' && this.props.label,
                'p-button-icon-right': position === 'right' && this.props.label,
                'p-button-icon-top': position === 'top' && this.props.label,
                'p-button-icon-bottom': position === 'bottom' && this.props.label
            });
            content = <span className={className}></span>;

            if (iconType !== 'string') {
                const defaultContentOptions = {
                    className,
                    element: content,
                    props: this.props
                };

                content = ObjectUtils.getJSXElement(icon, defaultContentOptions);
            }
        }

        return content;
    }

    renderLabel() {
        if (this.props.label) {
            return <span className="p-button-label p-c">{this.props.label}</span>;
        }

        return !this.props.children && !this.props.label && <span className="p-button-label p-c" dangerouslySetInnerHTML={{ __html: "&nbsp;" }}></span>
    }

    renderBadge() {
        if (this.props.badge) {
            const badgeClassName = classNames('p-badge', this.props.badgeClassName);

            return <span className={badgeClassName}>{this.props.badge}</span>
        }

        return null;
    }

    render() {
        let loadingOptions = this.getLoadingOptions();
        let { disabled, vertical, iconOnly } = this.getOptions(loadingOptions);
        let className = classNames('p-button p-component', this.props.className, {
            'p-button-icon-only': iconOnly,
            'p-button-vertical': vertical,
            'p-disabled': disabled
        });
        let loading = this.props.loading && this.renderIcon(loadingOptions.icon, loadingOptions.position);
        let icon = this.renderIcon(this.props.icon, this.props.iconPos);
        let label = this.renderLabel();
        let badge = this.renderBadge();

        let buttonProps = ObjectUtils.findDiffKeys(this.props, ButtonComponent.defaultProps);

        return (
            <button ref={(el) => this.getElementRef(el)} {...buttonProps} className={className} disabled={disabled}>
                {loading}
                {icon}
                {label}
                {this.props.children}
                {badge}
                <Ripple />
            </button>
        );
    }
}

export const Button = React.forwardRef((props, ref) => <ButtonComponent forwardRef={ref} {...props} />);
