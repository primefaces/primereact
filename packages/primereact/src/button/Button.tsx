'use client';
import { Component } from '@primereact/core/component';
import { useButton } from '@primereact/headless/button';
import { styles } from '@primereact/styles/button';
import { mergeProps } from '@primeuix/utils';
import { withComponent } from 'primereact/base';
import * as React from 'react';
import { ButtonProvider } from './Button.context';
import { defaultProps } from './Button.props';
import { ButtonGroup } from './group';

export const Button = withComponent({
    name: 'Button',
    defaultProps,
    styles,
    setup(instance) {
        const button = useButton(instance.inProps);

        return button;
    },
    render(instance) {
        const { id, props, ptmi, cx } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root')
            },
            ptmi('root')
        );

        return (
            <ButtonProvider value={instance}>
                <Component instance={instance} attrs={rootProps} children={props.children} />
            </ButtonProvider>
        );
    },
    components: {
        Group: ButtonGroup
    }
});
