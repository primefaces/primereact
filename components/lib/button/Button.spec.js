import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { snapshot } from '../../test';
import PrimeReact from '../api/PrimeReact';
import { Button } from './Button';

describe('Button', () => {
    snapshot(<Button label={'test'} visible={false} />, 'when visible is false Button return null');
    snapshot(<Button label={'test'} visible={true} />, 'when visible is true Button render correctly');
    snapshot(<Button label={'test'} iconPos={'bottom'} visible={true} />, 'when iconPos is bottom Button is vertical');
    snapshot(<Button visible={true} />, 'when label is empty it returns empty button');
    snapshot(<Button badge={'test'} />, 'when badge is true it renders Button with badge');
    snapshot(<Button />, 'when badge is null it renders Button without badge');
    snapshot(<Button loading={'test'} />, 'when click the button if loading is true it renders Button with loading icon');
    snapshot(<Button />, 'when click the button if loading is false it renders Button without loading icon');
    snapshot(<Button />, 'when label is true it renders Button with default aria label');
    snapshot(<Button label={'test'} />, 'when aria-label prop is not exist aria-label prop should be equal to label prop ');
    snapshot(<Button aria-label={'test'} />, 'when label prop is not exist label prop should be equal to aria-label prop');
    snapshot(<Button label={'test'} badge={'lost'} />, 'when using badge and label the aria-label should contain both values');
    snapshot(
        <>
            <Button label={'test'} badge={'lost'} />
        </>,
        'when using badge and label the aria-label should contain both values'
    );
    test('when using tooltip make sure the tooltip is rendered', async () => {
        // Arrange
        const { container } = render(<Button label={'test'} tooltip="Jest Tooltip" />);
        const button = container.getElementsByClassName('p-button')[0];
        const tooltipText = /Jest Tooltip/i;

        // tooltip does not exist to start
        expect(screen.queryByText(tooltipText)).toBeNull();

        // Act
        fireEvent.mouseEnter(button);

        // Assert
        await waitFor(() => screen.getByText(tooltipText));
        expect(screen.getByText(tooltipText)).toBeVisible();

        // tooltip disappears when we mouse out
        fireEvent.mouseLeave(button);
        expect(screen.queryByText(tooltipText)).toBeNull();
    });

    test('when button is clicked ensure onClick is fired', async () => {
        // Arrange
        const clickOn = jest.fn();
        const { container } = render(<Button onClick={clickOn} />);
        const button = container.getElementsByClassName('p-button')[0];

        // Act
        await userEvent.click(button);

        // Assert
        expect(button).toBeEnabled();
        expect(clickOn).toHaveBeenCalledTimes(1);
    });
    test('when button is disabled the click event should not fire', async () => {
        // Arrange
        const clickOn = jest.fn();
        const { container } = render(<Button onClick={clickOn} disabled />);
        const button = container.getElementsByClassName('p-button')[0];

        // Act
        await userEvent.click(button);

        // Assert
        expect(button).toBeDisabled();
        expect(clickOn).toHaveBeenCalledTimes(0);
    });
    test('when Ripple is enabled button should have ripple effect', async () => {
        // Arrange
        const clickOn = jest.fn();

        PrimeReact.ripple = true;
        const { container } = render(<Button onClick={clickOn} />);
        const button = container.getElementsByClassName('p-button')[0];

        // Act
        await userEvent.click(button);

        // Assert
        expect(button).toBeEnabled();
        expect(clickOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot();
    });
});
