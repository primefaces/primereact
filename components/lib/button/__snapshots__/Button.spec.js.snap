// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when Ripple is enabled button should have ripple effect 1`] = `
<div>
  <button
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
    <span
      class="p-ink p-ink-active"
      role="presentation"
      style="height: 0px; width: 0px;"
    />
  </button>
</div>
`;

exports[`Button when aria-label prop is not exist aria-label prop should be equal to label prop  1`] = `
<div>
  <button
    aria-label="test"
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
      test
    </span>
  </button>
</div>
`;

exports[`Button when badge is null it renders Button without badge 1`] = `
<div>
  <button
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
  </button>
</div>
`;

exports[`Button when badge is true it renders Button with badge 1`] = `
<div>
  <button
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
    <span
      class="p-badge"
    >
      test
    </span>
  </button>
</div>
`;

exports[`Button when click the button if loading is false it renders Button without loading icon 1`] = `
<div>
  <button
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
  </button>
</div>
`;

exports[`Button when click the button if loading is true it renders Button with loading icon 1`] = `
<div>
  <button
    class="p-button p-component p-button-icon-only p-disabled p-button-loading"
    disabled=""
  >
    <span
      class="p-button-icon p-c p-button-loading-icon pi pi-spinner pi-spin"
    />
    <span
      class="p-button-label p-c"
    >
       
    </span>
  </button>
</div>
`;

exports[`Button when iconPos is bottom Button is vertical 1`] = `
<div>
  <button
    aria-label="test"
    class="p-button p-component p-button-vertical"
  >
    <span
      class="p-button-label p-c"
    >
      test
    </span>
  </button>
</div>
`;

exports[`Button when label is empty it returns empty button 1`] = `
<div>
  <button
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
  </button>
</div>
`;

exports[`Button when label is true it renders Button with default aria label 1`] = `
<div>
  <button
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
  </button>
</div>
`;

exports[`Button when label prop is not exist label prop should be equal to aria-label prop 1`] = `
<div>
  <button
    aria-label="test"
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
       
    </span>
  </button>
</div>
`;

exports[`Button when using badge and label the aria-label should contain both values 1`] = `
<div>
  <button
    aria-label="test lost"
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
      test
    </span>
    <span
      class="p-badge"
    >
      lost
    </span>
  </button>
</div>
`;

exports[`Button when using badge and label the aria-label should contain both values 2`] = `
<div>
  <button
    aria-label="test lost"
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
      test
    </span>
    <span
      class="p-badge"
    >
      lost
    </span>
  </button>
</div>
`;

exports[`Button when visible is false Button return null 1`] = `<div />`;

exports[`Button when visible is true Button render correctly 1`] = `
<div>
  <button
    aria-label="test"
    class="p-button p-component"
  >
    <span
      class="p-button-label p-c"
    >
      test
    </span>
  </button>
</div>
`;
